magento/ece-tools
# Mcloud 5878: Create schema.error.yaml file with all errors and solutions

### Description

https://jira.corp.magento.com/browse/MCLOUD-5878
https://jira.corp.magento.com/browse/MCLOUD-5879

- Added schema.error.yaml file with all errors/warnings and solutions on how to fix the error.
- Added error codes for WARNING messages.
- Writes all errors and warnings in JSON format into separate log file `/var/log/cloud.error.log` with codes, and suggestions.

### Manual testing scenarios

1. Run deployment with few warnings in `/var/log/cloud.log`
2. Check that warnings messages contains error code 
```
WARNING: [2005] The following admin data is required to create an admin user during initial installation only and is ignored during the upgrade process: admin email, admin login
```
3. Check that file `/var/log/cloud.error.log` was created and it contains all warning/critical error messages (with error codes in cloud.log) in JSON format.
For example:
```
{"errorCode":2005,"stage":"deploy","step":"validate-config","title":"The following admin data is required to create an admin user during initial installation only and is ignored during upgrade process: admin email, admin login","type":"warning"}
{"errorCode":2007,"stage":"deploy","step":"validate-config","suggestion":"Change the version of PHP to the version that satisfies the restriction conditions.\nChange the PHP version in the .magento.app.yaml file or create a support ticket","title":"For Magento 2.3.5.0 recommended PHP version satisfying the constraint ~7.3.0. Currently installed PHP version 7.2.25","type":"warning"}
{"errorCode":132,"title":"Can't get version of elasticsearch: cURL error 7: Failed to connect to localhost port 9200: Connection refused (see https://curl.haxx.se/libcurl/c/libcurl-errors.html)"}
```
4. Run command `php vendor/bin/ece-tools error:show --json` 
and check that all errors from `/var/log/cloud.error.log` file are displayed in json format.
5. Run command `php vendor/bin/ece-tools error:show` 
and check that all errors from `/var/log/cloud.error.log` file are displayed in a human-readable format.

### Release notes

-  Added the `schema.error.yaml` file that includes all error and warning notifications that can occur during the Magento Cloud build, deploy, and post-deploy process along with suggestions for resolving the errors.

-  Updated the Cloud error log (/var/log/cloud.error.log) to provide entries in JSON format to make the log easier to parse programmatically. See [Error message reference for ece-tools](https://devdocs.magento.com/cloud/reference/error-codes.html).


### Associated documentation updates

https://github.com/magento/devdocs/pull/7578
https://github.com/magento/devdocs/pull/7573

### Contribution checklist
